src/drivers/wa-sqlite/adapter.ts

```diff
@@ -0,0 +1,180 @@
+import {
+  DatabaseAdapter as DatabaseAdapterInterface,
+  RunResult,
+  Transaction as Tx,
+} from '../../electric/adapter'
+import { Database } from './database'
+import { parseTableNames, QualifiedTablename, Row, Statement } from '../../util'
+import { resultToRows } from '../absurd-sql'
+import { isInsertUpdateOrDeleteStatement } from '../../util/statements'
+import { Mutex } from 'async-mutex'
+
+export class DatabaseAdapter implements DatabaseAdapterInterface {
+  db: Database
+  txMutex: Mutex
+
+  constructor(db: Database) {
+    this.db = db
+    this.txMutex = new Mutex()
+  }
+
+  async runInTransaction(...statements: Statement[]): Promise<RunResult> {
+    // Uses a mutex to ensure that transactions are not interleaved.
+    // This is needed because transactions are executed manually using `BEGIN` and `COMMIT` commands.
+    const release = await this.txMutex.acquire()
+    let open = false
+    let rowsAffected = 0
+    try {
+      await this.db.exec({ sql: 'BEGIN' })
+      open = true
+      for (const stmt of statements) {
+        await this.db.exec(stmt)
+        if (isInsertUpdateOrDeleteStatement(stmt.sql)) {
+          // Fetch the number of rows affected by the last insert, update, or delete
+          rowsAffected += await this.db.getRowsModified()
+        }
+      }
+      return {
+        rowsAffected: rowsAffected,
+      }
+    } catch (error) {
+      await this.db.exec({ sql: 'ROLLBACK' })
+      open = false
+      throw error // rejects the promise with the reason for the rollback
+    } finally {
+      if (open) {
+        await this.db.exec({ sql: 'COMMIT' })
+      }
+      release()
+    }
+  }
+
+  async transaction<T>(
+    f: (_tx: Tx, setResult: (res: T) => void) => void
+  ): Promise<T | void> {
+    const release = await this.txMutex.acquire()
+
+    try {
+      await this.db.exec({ sql: 'BEGIN' })
+    } catch (e) {
+      release()
+      throw e
+    }
+
+    return new Promise((resolve, reject) => {
+      const releaseAndReject = (err?: any) => {
+        // if the tx is rolled back, release the lock and reject the promise
+        release()
+        reject(err)
+      }
+
+      const tx = new Transaction(this, releaseAndReject)
+
+      f(tx, (res) => {
+        // Commit the transaction when the user sets the result.
+        // This assumes that the user does not execute any more queries after setting the result.
+        this.db
+          .exec({ sql: 'COMMIT' })
+          .then(() => {
+            release()
+            resolve(res)
+          })
+          .catch((err) => {
+            release()
```

Comment:
Indeed, i believe that if `release()` is called in a `finally` block, then the reaction to `resolve()`/`reject()` may be triggered before the mutex is released. So if the reaction tries to execute a new transaction it will have to wait. This should not really cause any problems though. Do you prefer to move `release()` to a finally block?

Reply:
No, like you did is correct. I think that the final is 'almost a good interface' :D but doesn't solve all code redundancies: the ideal behaviour would be if the resolve/reject would be aware of the final and only effectively return after running final (maybe that is impossible).

Honestly I couldn't conclude the behaviour from the docs: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally

I wonder if @icehaunter knows best.